File: dexfile/src/main/java/com/github/netomi/bat/dexfile/annotation/ParameterAnnotation.java
/*
 *  Copyright (c) 2020 Thomas Neidhart.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.github.netomi.bat.dexfile.annotation;

import com.github.netomi.bat.dexfile.*;
import com.github.netomi.bat.dexfile.io.DexDataInput;
import com.github.netomi.bat.dexfile.io.DexDataOutput;
import com.github.netomi.bat.dexfile.visitor.AnnotationSetVisitor;
import com.github.netomi.bat.dexfile.visitor.AnnotationVisitor;
import com.github.netomi.bat.dexfile.visitor.DataItemVisitor;

import java.util.Objects;

import static com.github.netomi.bat.dexfile.DexConstantsKt.NO_INDEX;

/**
 * A class representing a parameter annotation format inside a dex file.
 *
 * @see <a href="https://source.android.com/devices/tech/dalvik/dex-format#parameter-annotation">parameter annotation format @ dex format</a>
 */
public class ParameterAnnotation
extends      DexContent
{
    private int methodIndex;       // uint
    private int annotationsOffset; // uint

    private AnnotationSetRefList annotationSetRefList;

    public static ParameterAnnotation of(int methodIndex, AnnotationSetRefList annotationSetRefList) {
        return new ParameterAnnotation(methodIndex, annotationSetRefList);
    }

    public static ParameterAnnotation readContent(DexDataInput input) {
        ParameterAnnotation parameterAnnotation = new ParameterAnnotation();
        parameterAnnotation.read(input);
        return parameterAnnotation;
    }

    private ParameterAnnotation() {
        this(NO_INDEX, null);
    }

    private ParameterAnnotation(int methodIndex, AnnotationSetRefList annotationSetRefList) {
        this.methodIndex          = methodIndex;
        this.annotationSetRefList = annotationSetRefList;
    }

    public int getMethodIndex() {
        return methodIndex;
    }

    public int getAnnotationsOffset() {
        return annotationsOffset;
    }

    public MethodID getMethodID(DexFile dexFile) {
        return dexFile.getMethodID(methodIndex);
    }

    @Override
    protected void read(DexDataInput input) {
        methodIndex       = input.readInt();
        annotationsOffset = input.readInt();
    }

    @Override
    public void readLinkedDataItems(DexDataInput input) {
        input.setOffset(annotationsOffset);
        annotationSetRefList = AnnotationSetRefList.readContent(input);
    }

    @Override
    public void updateOffsets(DataItem.Map dataItemMap) {
        annotationsOffset = dataItemMap.getOffset(annotationSetRefList);
    }

    @Override
    public void write(DexDataOutput output) {
        output.writeInt(methodIndex);
        output.writeInt(annotationsOffset);
    }

    public void accept(DexFile dexFile, ClassDef classDef, AnnotationSetVisitor visitor) {
        visitor.visitParameterAnnotationSet(dexFile, classDef, this, annotationSetRefList);
    }

    public void accept(DexFile dexFile, ClassDef classDef, int parameterIndex, AnnotationVisitor visitor) {
        if (parameterIndex >= 0 && parameterIndex < annotationSetRefList.getAnnotationSetRefCount()) {
            AnnotationSetRef annotationSetRef = annotationSetRefList.getAnnotationSetRef(parameterIndex);
            if (annotationSetRef != null) {
                annotationSetRef.accept(dexFile, classDef, visitor);
            }
        }
    }

    @Override
    public void dataItemsAccept(DexFile dexFile, DataItemVisitor visitor) {
        visitor.visitParameterAnnotations(dexFile, this, annotationSetRefList);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        ParameterAnnotation other = (ParameterAnnotation) o;
        return methodIndex == other.methodIndex &&
               Objects.equals(annotationSetRefList, other.annotationSetRefList);
    }

    @Override
    public int hashCode() {
        return Objects.hash(methodIndex, annotationSetRefList);
    }

    @Override
    public String toString() {
        return String.format("ParameterAnnotation[methodIndex=%d]", methodIndex);
    }
}

File: dexfile/src/main/java/com/github/netomi/bat/dexfile/annotation/ParameterAnnotation.kt
/*
 *  Copyright (c) 2020 Thomas Neidhart.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.github.netomi.bat.dexfile.annotation

import com.github.netomi.bat.dexfile.*
import com.github.netomi.bat.dexfile.annotation.visitor.AnnotationSetVisitor
import com.github.netomi.bat.dexfile.annotation.visitor.AnnotationVisitor
import com.github.netomi.bat.dexfile.annotation.visitor.allAnnotations
import com.github.netomi.bat.dexfile.io.DexDataInput
import com.github.netomi.bat.dexfile.io.DexDataOutput
import com.github.netomi.bat.dexfile.visitor.*
import com.github.netomi.bat.dexfile.visitor.PropertyAccessor
import com.github.netomi.bat.dexfile.visitor.ReferencedIDVisitor
import java.util.*

/**
 * A class representing a parameter annotation format inside a dex file.
 *
 * @see [parameter annotation format @ dex format](https://source.android.com/devices/tech/dalvik/dex-format.parameter-annotation)
 */
class ParameterAnnotation private constructor(methodIndex:          Int                  = NO_INDEX,
                                              annotationSetRefList: AnnotationSetRefList = AnnotationSetRefList.empty()) : DexContent() {

    var annotationsOffset = 0
        private set

    var methodIndex: Int = methodIndex
        private set

    internal var annotationSetRefList: AnnotationSetRefList = annotationSetRefList
        private set

    fun getMethodID(dexFile: DexFile): MethodID {
        return dexFile.getMethodID(methodIndex)
    }

    override fun read(input: DexDataInput) {
        methodIndex       = input.readInt()
        annotationsOffset = input.readInt()
    }

    override fun readLinkedDataItems(input: DexDataInput) {
        input.offset = annotationsOffset
        annotationSetRefList = AnnotationSetRefList.read(input)
    }

    override fun updateOffsets(dataItemMap: DataItem.Map) {
        annotationsOffset = dataItemMap.getOffset(annotationSetRefList)
    }

    override fun write(output: DexDataOutput) {
        output.writeInt(methodIndex)
        output.writeInt(annotationsOffset)
    }

    fun accept(dexFile: DexFile, classDef: ClassDef, visitor: AnnotationSetVisitor) {
        visitor.visitParameterAnnotationSetRefList(dexFile, classDef, this, annotationSetRefList)
    }

    fun accept(dexFile: DexFile, classDef: ClassDef, parameterIndex: Int, visitor: AnnotationSetVisitor) {
        if (parameterIndex in 0 until annotationSetRefList.annotationSetRefCount) {
            val annotationSetRef = annotationSetRefList.getAnnotationSetRef(parameterIndex)
            visitor.visitParameterAnnotationSet(dexFile, classDef, this, parameterIndex, annotationSetRef.annotationSet)
        }
    }

    fun annotationsAccept(dexFile: DexFile, classDef: ClassDef, parameterIndex: Int, visitor: AnnotationVisitor) {
        accept(dexFile, classDef, parameterIndex, allAnnotations(visitor))
    }

    override fun dataItemsAccept(dexFile: DexFile, visitor: DataItemVisitor) {
        visitor.visitParameterAnnotations(dexFile, this, annotationSetRefList)
        annotationSetRefList.dataItemsAccept(dexFile, visitor)
    }

    internal fun referencedIDsAccept(dexFile: DexFile, visitor: ReferencedIDVisitor) {
        visitor.visitMethodID(dexFile, PropertyAccessor({ methodIndex }, { methodIndex = it }))
        for (annotationSetRef in annotationSetRefList.annotationSetRefs) {
            annotationSetRef.referencedIDsAccept(dexFile, visitor)
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false

        val o = other as ParameterAnnotation

        return methodIndex          == o.methodIndex &&
               annotationSetRefList == o.annotationSetRefList
    }

    override fun hashCode(): Int {
        return Objects.hash(methodIndex, annotationSetRefList)
    }

    override fun toString(): String {
        return "ParameterAnnotation[methodIndex=${methodIndex},annotationSetRefList=${annotationSetRefList.annotationSetRefCount} items]"
    }

    companion object {
        fun of(methodIndex: Int, annotationSetRefList: AnnotationSetRefList): ParameterAnnotation {
            return ParameterAnnotation(methodIndex, annotationSetRefList)
        }

        internal fun read(input: DexDataInput): ParameterAnnotation {
            val parameterAnnotation = ParameterAnnotation()
            parameterAnnotation.read(input)
            return parameterAnnotation
        }
    }
}